src/daft-scan/src/scan_task_iters/split_parquet_files_by_rowgroup.rs

```diff
@@ -0,0 +1,465 @@
+enum MaybeSplitScanTaskByParquetRowgroups {
+    NoSplit(ScanTaskRef),
+    Split(SplittableScanTaskRef),
+}
+
+struct SplittableScanTaskRef(ScanTaskRef);
+
+impl SplittableScanTaskRef {
+    /// Uses Parquet FileMetaData to create an Iterator of ScanTasks that are the result of splitting the existing Parquet ScanTask
+    pub fn get_scan_task_iter<'a>(
+        &'a self,
+        file_metadata: FileMetaData,
+        config: &'a DaftExecutionConfig,
+    ) -> ParquetFileSplitter<'a> {
+        ParquetFileSplitter::new(self.0.as_ref(), file_metadata, config)
+    }
+
+    pub fn get_path(&self) -> &str {
+        match &self.0.sources[..] {
+            [data_source] => data_source.get_path(),
+            _ => panic!("Should not have more than 1 source in a SplitScanTaskRef"),
+        }
+    }
+
+    pub fn get_parquet_source_config(&self) -> &ParquetSourceConfig {
+        if let FileFormatConfig::Parquet(cfg @ ParquetSourceConfig { .. }) =
+            self.0.file_format_config.as_ref()
+        {
+            cfg
+        } else {
+            panic!("SplitScanTaskRef should be a Parquet file");
+        }
+    }
+}
+
+impl MaybeSplitScanTaskByParquetRowgroups {
+    /// Optionally creates [`SplitScanTaskInfo`] depending on whether or not a given ScanTask can be split
+    pub fn from_scan_task(scan_task: ScanTaskRef, config: &DaftExecutionConfig) -> Self {
+        /* Only split parquet tasks if they:
+            - have one source
+            - use native storage config
+            - have no specified chunk spec or number of rows
+            - have size past split threshold
+            - no iceberg delete files
+        */
+        if let (
+            FileFormatConfig::Parquet(ParquetSourceConfig { .. }),
+            StorageConfig::Native(_),
+            [source],
+            Some(None),
+            None,
+            est_materialized_size,
+        ) = (
+            scan_task.file_format_config.as_ref(),
+            scan_task.storage_config.as_ref(),
+            &scan_task.sources[..],
+            scan_task.sources.first().map(DataSource::get_chunk_spec),
+            scan_task.pushdowns.limit,
+            scan_task.estimate_in_memory_size_bytes(Some(config)),
+        ) && est_materialized_size.map_or(true, |est| est > config.scan_tasks_max_size_bytes)
+            && source
+                .get_iceberg_delete_files()
+                .map_or(true, std::vec::Vec::is_empty)
+        {
+            Self::Split(SplittableScanTaskRef(scan_task))
+        } else {
+            Self::NoSplit(scan_task)
+        }
+    }
+
+    pub fn unwrap(self) -> ScanTaskRef {
+        match self {
+            Self::NoSplit(scan_task) => scan_task,
+            Self::Split(split_scan_task) => split_scan_task.0,
+        }
+    }
+}
+
+/// Helper method to split ScanTasks using [`SplitScanTaskInfo`]
+///
+/// Returns a HashMap of {idx: split_results}
+fn split_scan_tasks_by_parquet_metadata(
```

Comment:
this is only called by 1 method. this should be nested and hidden by that struct